pwiz_tools/Skyline/TestUtil/TestFunctional.cs

```diff
@@ -1089,47 +1087,48 @@ private void OnAuditLogEntryAdded(object sender, AuditLogEntry.AuditLogEntryAdde
             }
 
-            var toFile = GetLogFilePath(AuditLogTutorialDir);
+            // Ensure expected tutorial log file exists unless recording
+            var projectFile = GetLogFilePath(AuditLogTutorialDir);
+            bool existsInProject = File.Exists(projectFile);
+            if (!IsRecordAuditLogForTutorials)
+            {
+                Assert.IsTrue(existsInProject,
+                    "Log file for test \"{0}\" does not exist, set IsRecordAuditLogForTutorials=true to create it",
+                    TestContext.TestName);
+            }
 
-            // If nothing has changed skip copying and failing
-            var expected = File.ReadAllText(toFile);
-            var actual = File.ReadAllText(fromFile);
+            // Compare file contents
+            var expected = existsInProject ? File.ReadAllText(projectFile) : string.Empty;
+            var actual = File.ReadAllText(recordedFile);
             if (Equals(expected, actual))
+            {
+                Helpers.TryTwice(() => File.Delete(recordedFile));    // Avoid appending to the same file on multiple runs
                 return;
+            }
 
-            File.Copy(fromFile, toFile, true);
-            Helpers.TryTwice(() => File.Delete(fromFile));    // Avoid appending to the same file on multiple runs
-            Assert.Fail("Successfully recorded tutorial audit log");
-        }
-
-        private void VerifyAuditLogCorrect()
-        {
-            if (!AuditLogCompareLogs || IsRecordAuditLogForTutorials)
-                return;
-
-            // Ensure expected tutorial log file exists
-            var expectedFilePath = GetLogFilePath(AuditLogTutorialDir);
-            Assert.IsTrue(File.Exists(expectedFilePath),
-                "Log file for test \"{0}\" does not exist, set IsRecordAuditLogForTutorials=true to create it",
-                TestContext.TestName);
-
-            // Compare file contents
-            var expected = File.ReadAllText(expectedFilePath);
-            var actualFilePath = GetLogFilePath(AuditLogDir);
-            var actual = File.ReadAllText(actualFilePath);
-            AssertEx.NoDiff(expected, actual);
-            Helpers.TryTwice(() => File.Delete(actualFilePath));    // Avoid appending to the same file on multiple runs
+            // They are not equal. So, report an intelligible error and potentially copy
+            // a new expected file to the project if in record mode.
+            if (!IsRecordAuditLogForTutorials)
+                AssertEx.NoDiff(expected, actual);
+            else
+            {
+                // Copy the just recorded file to the project for comparison or commit
+                File.Copy(recordedFile, projectFile, true);
+                if (!existsInProject)
+                    Assert.Fail("Successfully recorded tutorial audit log");
+                else
+                    AssertEx.NoDiff(expected, actual, "Successfully recorded changed tutorial audit log:");

```

Comment:
This means that if someone records the tutorial logs and there is no difference, the tutorial test won't fail and it's much more likely that we push the code with IsRecordAuditLogForTutorials set to true. Maybe we should always fail if IsRecordAuditLogForTutorials is true?

Reply:
Yeah. I knew that, but the other way makes it much more manual to get everything updated in SkylineTester. But, okay, I will rethink my solution to that to only skip failing in the case of a Quality pass. That will probably allow me to be even more fully-automatic for a complete update and still fail in more common cases that will keep us from committing in record mode.

Note that committing in record mode still fails if the anything changes in my current solution. So, it doesn't mean the tests will pass on bad results. It just means you can end up copying files into your project when the test does fail.

Reply:
I actually felt that the original implementation was part of the problem that caused you to merge before everything actually passed running under SkylineTester conditions, because it required you to run under single-language conditions to record everything and then failed under multi-language conditions.

Reply:
Yes, and it made recording logs for multiple languages much more tedious. This makes me think that there should be a way for us to still run a test in other languages even if it fails in another.

Reply:
We could make it dependent on Program.SkylineOffscreen. That would allow you to test in record mode, as I have been doing, using the Quality tab, and getting everything recorded in the process, but Skyline would still fail enough to provide a good balance between keeping what happened to you from happening again and keeping people from committing with the flag on.

Reply:
Or, yeah, a way to control when TestRunner stops running a test (either at first failure) or after running all languages. That would probably be better. I will leave this for you, after I merge this to get nightly tests working again.

Reply:
Okay sounds good